reactive/kubernetes_master.py

```diff
@@ -2575,3 +2576,77 @@ def add_systemd_iptables_patch():
 
     # enable and run the service
     service_resume(service_name)
+
+
+@when('leadership.is_leader',
+      'kubernetes-master.components.started',
+      'endpoint.prometheus.joined',
+      'certificates.ca.available')
+def register_prometheus_jobs():
+    prometheus = endpoint_from_flag('endpoint.prometheus.joined')
+    # tls = endpoint_from_flag('certificates.ca.available')
+    loadbalancer = endpoint_from_flag('loadbalancer.available')
+    forced_lb_ips = hookenv.config('loadbalancer-ips').split()
+    hacluster_vip = get_hacluster_ip_or_hostname()
+
+    if forced_lb_ips:
+        # if the user gave us IPs for the load balancer, assume they know
+        # what they are talking about and use that instead of our information.
+        round_robin = get_unit_number() % len(forced_lb_ips)
+        k8s_api_endpoint = forced_lb_ips[round_robin]
+    elif hacluster_vip:
+        k8s_api_endpoint = hacluster_vip
+    elif loadbalancer:
+        lb_addresses = loadbalancer.get_addresses_ports()
+        k8s_api_endpoint = lb_addresses[0].get('public-address')
+    else:
+        k8s_api_endpoint = prometheus.ingress_address
+    k8s_password = get_password('basic_auth.csv', 'admin')
+
+    templates_dir = Path('templates')
+    for job_file in Path('templates/prometheus').glob('*.yaml.j2'):
+        prometheus.register_job(
+            job_name=job_file.name.split('.')[0],
+            job_data=yaml.safe_load(
+                render(source=job_file.relative_to(templates_dir),
+                       target=None,  # don't write file, just return rendered
+                       context={
+                           'k8s_api_endpoint': k8s_api_endpoint,
+                           'k8s_password': k8s_password,
+                       })
+            ),
+        )
+
+
+def detect_telegraf():
+    # Telegraf uses the implicit juju-info relation, which makes it difficult
+    # to tell if it's related. The "best" option is to look for the subordinate
+    # charm on disk.
```

Comment:
On DISK? This seems very brittle. :disappointed:

Reply:
Yeah, it does.  I had discussed this with @kwmonroe and we couldn't come up with a better solution, but I'd really love to find one, if you have some suggestions.  Options we considered:

* Make the dashboard unconditional (the deployment might not even use Telegraf)
* Make the dashboard conditional on the presence of Prometheus (could use Prometheus w/o Telegraf)
* Make the dashboard conditional on `which telegraf` (still requires on-disk checks and is really just brittle in a different way)

Thrown into the mix is that the Grafana charm includes a Telegraf dashboard out of the box which has some overlap, but which apparently [might not always be loaded](https://github.com/conjure-up/spells/blob/master/charmed-kubernetes/addons/prometheus/steps/01_install-prometheus/after-deploy#L28-L31) when we want it to be.

Reply:
:crying_cat_face: